### PR TITLE
single client with futures api

### PR DESCRIPTION
**Pull request by jasonahills**
Changes api to a single client which returns `Stream`s and `Future`s on subscribe and publish respectively.

### CODE REVIEW

File: examples/receiver.rs [hunk 1]
```diff
@@ -26,21 +26,25 @@ struct Stuff {
...
+    subscribe_key: opt.subscribe_key,
+    client_uuid: opt.client_uuid,
+  });
 
-  tokio::run(rx.for_each(|v| {
-    println!("whoa! a message! {:?}", v);
+  let channel = opt.channel;
+  let group = opt.group;
+
+  let task = client.unwrap().subscribe::<Stuff>(&channel, &group).unwrap();
```
Comment (Xaeroxe) [inline examples/receiver.rs]: Nitpick: It might be better to move the `client.unwrap()` to the place where the client is constructed, especially since `subscribe` doesn't consume the client.

File: examples/receiver.rs [hunk 2]
```diff
@@ -26,21 +26,25 @@ struct Stuff {
...
+  tokio::run(task.for_each(|v| {
+    match v {
+      Ok(i) => println!("whoa! a message! {:?}", i),
+      Err(e) => println!("error {:?}", e),
+    }
     Ok(())
-  }))
+  }));
+
+  println!("how did we get here?");
```
Comment (Xaeroxe) [inline examples/receiver.rs]: I might even go so far as to make this a `panic`.

File: examples/sender.rs [hunk 3]
```diff
@@ -26,23 +26,29 @@ struct Stuff {
 
 fn main() {
   let opt = Opt::from_args();
-  let mut client = PublishClient::new(ClientConfig {
+  let client = Client::new(ClientConfig {
```
Comment (Xaeroxe) [inline examples/sender.rs:29]: Always happy to see a `mut` go away :smile:

File: src/lib.rs [hunk 4]
```diff
@@ -3,38 +3,160 @@ use futures::sync::mpsc::{Receiver, Sender};
...
+  group: CString,
+}
+
 struct SubscribeUserData<T> {
   channel: CString,
   tx: Sender<Result<T, ClientError>>,
 }
 
-pub struct SubscribeClient<T> {
+pub struct Client {
```
Comment (Xaeroxe) [inline src/lib.rs:33]: Nitpick: Public structures should, if possible, derive the [most common std traits](https://doc.rust-lang.org/rust-by-example/trait/derive.html). Doing this aggressively prevents needing commits like https://github.com/vivint-smarthome/zugzug/commit/2d7cbe6ff399f6c5307f029d86333ea112636857

File: src/lib.rs [hunk 5]
```diff
@@ -3,38 +3,160 @@ use futures::sync::mpsc::{Receiver, Sender};
...
+  auth_key: CString,
+  publish_key: CString,
+  subscribe_key: CString,
+  client_uuid: CString,
+}
+
+impl Client {
+  pub fn new(config: ClientConfig) -> Result<Self, std::ffi::NulError> {
```
Comment (Xaeroxe) [inline src/lib.rs]: So on further review, this actually can't fail so it should just return `Self`. The same is true for other places that we're creating `CString`s.

The reason this can't fail, is that no valid UTF-8 character contains Nul, and all rust Strings are valid UTF-8. [CString::new](https://doc.rust-lang.org/stable/std/ffi/struct.CString.html#method.new) actually takes a `Into<Vec<u8>>` which does not provide this guarantee. However, since we're passing a structure that does provide this guarantee, this can't fail. Consequently when building a `CString` from `String`s we should always `.unwrap()`, leading this to be a function that can never fail.

File: src/lib.rs [hunk 6]
```diff
@@ -3,38 +3,160 @@ use futures::sync::mpsc::{Receiver, Sender};
...
+      client_uuid,
+    })
+  }
+
+  pub fn subscribe<'a, T: Send + Sync + Deserialize<'a>>(
+    &self,
+    channel: &str,
+    group: &str,
+  ) -> Result<Subscription<T>, std::ffi::NulError> {
+    let channel_c = CString::new(channel)?;
```
Comment (Xaeroxe) [inline src/lib.rs]: This also can't fail

File: src/lib.rs [hunk 7]
```diff
@@ -3,38 +3,160 @@ use futures::sync::mpsc::{Receiver, Sender};
...
+
+    Ok(Subscription::new(config))
+  }
+
+  pub fn publish<T: Serialize>(
+    &self,
+    channel: &str,
+    group: &str,
+    body: T,
+  ) -> Result<PublishFuture, std::ffi::NulError> {
```
Comment (Xaeroxe) [inline src/lib.rs]: Can't fail

File: src/lib.rs [hunk 8]
```diff
@@ -3,38 +3,160 @@ use futures::sync::mpsc::{Receiver, Sender};
...
   _auth_key: CString,
   _publish_key: CString,
   _subscribe_key: CString,
-  _group: CString,
   _client_uuid: CString,
+  _channel: CString,
+  _group: CString,
+}
+
+// The pointers in `Subscription` are thread-safe, so we can implement this.
```
Comment (Xaeroxe) [inline src/lib.rs:109]: If possible, it'd be neat if you could provide a citation on this statement

File: zugzug-sys/build.rs [hunk 9]
```diff
@@ -65,6 +65,7 @@ fn main() {
     .clang_arg(format!("-I{}", upstream_build_dir.display()))
     .clang_arg(format!("-I{}", upstream_build_dir_posix.display()))
     .clang_arg("-DPUBNUB_CALLBACK_API=1")
+    .clang_arg("-DPUBNUB_THREADSAFE=1") // Makes contexts thread-safe, justifying our making them Send and Sync.
```
Comment (Xaeroxe) [inline zugzug-sys/build.rs:76]: I wonder what the drawback to having this enabled is. Not to say we shouldn't do it, I just find their choice of using a preprocessor definition to be odd.

**Review by Xaeroxe — CHANGES_REQUESTED** [review]
This is awesome! Thanks! I spotted a few places we can make it even better. Let me know what you think.

File: src/lib.rs [hunk 10]
```diff
@@ -205,22 +243,14 @@ pub struct PublishFuture {
...
     let ctx = unsafe {
       let ctx = pubnub_alloc();
-      pubnub_init(ctx, publish_key.as_ptr(), subscribe_key.as_ptr());
-      pubnub_set_uuid(ctx, client_uuid.as_ptr());
-      pubnub_set_auth(ctx, auth_key.as_ptr());
+      pubnub_init(ctx, config.publish_key.as_ptr(), config.subscribe_key.as_ptr());
+      pubnub_set_uuid(ctx, config.client_uuid.as_ptr());
+      pubnub_set_auth(ctx, config.auth_key.as_ptr());
       ctx
     };
```
Comment (ErichDonGubler) [inline src/lib.rs]: You could get a far smaller diff (and be more resistant to changes to `config`'s structure!) by just destructuring thusly:

```rust
let Config {
    msg,
    channel,
    auth_key,
    publish_key,
    subscribe_key,
    group,
    client_uuid,
} = config;
```

File: src/lib.rs [hunk 11]
```diff
@@ -3,38 +3,160 @@ use futures::sync::mpsc::{Receiver, Sender};
...
+      channel: channel.clone(), // TODO: can this just be a reference?
+    }));
+
+    let ctx = unsafe {
+      let ctx = pubnub_alloc();
+      pubnub_init(ctx, publish_key.as_ptr(), subscribe_key.as_ptr());
+      pubnub_set_uuid(ctx, client_uuid.as_ptr());
+      pubnub_set_auth(ctx, auth_key.as_ptr());
+      pubnub_register_callback(ctx, Some(subscribe_callback::<T>), user_data as *mut std::ffi::c_void);
+      pubnub_subscribe(ctx, channel.as_ptr(), std::ptr::null()); // TODO: technically we shouldn't call this line until the stream gets polled the first time.
```
Comment (ErichDonGubler) [inline src/lib.rs:137]: > `// TODO: technically we shouldn't call this line until the stream gets polled the first time.`

Really? This is surprising to me. What's the reasoning here? I probably just need to learn the domain better. :)

File: src/lib.rs (hunk 5 above)
Comment (jasonahills) [reply src/lib.rs]: That's awesome!

File: src/lib.rs (hunk 8 above)
Comment (ErichDonGubler) [reply src/lib.rs:109]: This was implemented [here](https://github.com/vivint-smarthome/zugzug/pull/3/files/d71ff6d45da181935cb73361d2eb9b805cf0aafa#diff-2d0141ae178b6796a956c21fb25b2c45R68) (the `-DPUBNUB_THREADSAFE=1` arg for compilation). Maybe just add a note referencing that?

File: src/lib.rs (hunk 11 above)
Comment (jasonahills) [reply src/lib.rs:137]: Maybe I'm misunderstanding the domain, but my understanding is that for a `Future`/`Stream`, work should be driven by `poll`.  I would expect to be able to generate `Future`s as I please and only really pay for them (in this case with network I/O) when I start running them.

Again though, I would actually defer to your understanding in this domain.

File: zugzug-sys/build.rs (hunk 9 above)
Comment (ErichDonGubler) [reply zugzug-sys/build.rs:76]: [Relevant upstream docs](https://www.pubnub.com/docs/posix-c/api-reference-configuration#Thread_safety):

> #### [Thread safety](https://www.pubnub.com/docs/posix-c/api-reference-configuration#Thread_safety)
> 
> C-core supports thread-safe operation, though, for performance, you may think about not using it. To use thread-safety support, define the preprocessor symbol `PUBNUB_THREADSAFE` (just define it, the value does not matter).
> 
> ##### [Thread-safe Usage](https://www.pubnub.com/docs/posix-c/api-reference-configuration#thread-safe-usage)
> 
> Thread safety is internal. Just because you can access the pubnub context through the Pubnub C-core SDK API from different threads safely, doesn't mean you're `off the hook` for your own data that is related to a context. For example, if you're using the callback interface and signalling an event from it to other (`worker`) thread(s), you have to synchronise that data transfer yourself.
> 
> If you compiled thread-safety support in, you are free to access the same context from different threads, pretty much in any way you wish. However, there are some advised guidelines you should follow:


TL;DR: it's similar to the difference between `Arc` vs `Rc`. In fact, I'd be surprised if it wasn't for the exact same reasons.

**Review by ErichDonGubler — APPROVED** [review]
I'll defer to @Xaeroxe's request for changes before approving, but otherwise LGTM...assuming you've got your C contract right. Want me to review that?

EDIT: And I hit "Approve" by accident. Heh.

File: src/lib.rs (hunk 4 above)
Comment (jasonahills) [reply src/lib.rs:33]: Definitely.  I've made the change here.  Are there any traits that would be appropriate to derive for the futures and streams in our API?

File: src/lib.rs (hunk 8 above)
Comment (jasonahills) [reply src/lib.rs:109]: I'm adding a link to this: https://www.pubnub.com/docs/posix-c/api-reference-configuration#Thread_safety

File: zugzug-sys/build.rs (hunk 9 above)
Comment (ErichDonGubler) [reply zugzug-sys/build.rs:76]: If we were determined to make this really nice for everybody else that may use this, we would probably want to gate it behind a feature that's on by default (i.e., `thread-safe`). Either way, I think this can be resolved?

Comment (jasonahills) [reply zugzug-sys/build.rs:76]: Yeah.  I'm somewhat new to this space, but I too was surprised with how much they configure that way.

**Comment by jasonahills** [comment]
> I'll defer to @Xaeroxe's request for changes before approving, but otherwise LGTM...assuming you've got your C contract right. Want me to review that?
> 
> EDIT: And I hit "Approve" by accident. Heh.

The C contract is going to be the area where I will be weakest.  I'm happy for a review of that.

File: src/lib.rs (hunk 5 above)
Comment (Xaeroxe) [reply src/lib.rs]: So I need to back off this statement slightly, because further investigation revealed failure is technically possible.

Null is not contained in any valid UTF-8 character, except for the Null character itself, which is valid UTF-8. Rust does not have Null terminated strings, so this character is technically part of the character set a string can contain. See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5989145edbf21d6bff7dbed4f0d7b9a9).

However, Null has only ever had one purpose, and that is to indicate the end of a string. As far as I can tell it's never been used for anything else. So, the assumption that the config strings do not contain Null, is imo a very reasonable one, and not something worth exposing to the library user as a case they might need to handle.

File: src/lib.rs (hunk 10 above)
Comment (jasonahills) [reply src/lib.rs]: Hmm, I'm generally a huge fan of destructuring, but I'm not seeing the benefits here (I'd love to hear more from your perspective though).  It seems like without the destructure, any change to the name of a field in the `Config` would require exactly one change in the body of the function.  But with the destructure it would require two.

Additionally, the destructure adds a little boilerplate/noise in the body of the function.

Again, I'm not seeing the benefit of the change yet, but I'd love to hear more about why you do.

File: src/lib.rs (hunk 5 above)
Comment (ErichDonGubler) [reply src/lib.rs]: Just for reference, here's [the relevant `std::ffi::CString` section](https://doc.rust-lang.org/std/ffi/struct.CString.html#creating-a-cstring):

> #### [Creating a `CString`](https://doc.rust-lang.org/std/ffi/struct.CString.html#creating-a-cstring)
> 
> A `CString` is created from either a byte slice or a byte vector, or anything that implements [`Into`](https://doc.rust-lang.org/std/convert/trait.Into.html)`<`[`Vec`](https://doc.rust-lang.org/std/vec/struct.Vec.html)`<`[`u8`](https://doc.rust-lang.org/std/primitive.u8.html)`>>` (for example, you can build a `CString` straight out of a [`String`](https://doc.rust-lang.org/std/string/struct.String.html) or a [`&str`](https://doc.rust-lang.org/std/primitive.str.html), since both implement that trait).
> 
> The [`new`](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.new) method will actually check that the provided `&[u8]` does not have 0 bytes in the middle, and return an error if it finds one.

File: src/lib.rs (hunk 10 above)
Comment (ErichDonGubler) [reply src/lib.rs]: You're right in that it would create two places where the binding to a field is referenced, and it would be an additional statement before your logic. I guess I don't mind that when placed alongside the benefit of refactor resilience -- it can move more errors to compile time, specifically for the case when you add a field you intend to use in a constructor like this.

Comment (jasonahills) [reply src/lib.rs]: That makes sense.  So the refactor resilience is that if I add something to the struct, I'm forced to use it.  Cool.  I'll make the change.

File: src/lib.rs (hunk 4 above)
Comment (Xaeroxe) [reply src/lib.rs:33]: Probably not, and the reason I say that is that those structures represent an interface, rather than data. This looks great!

**Review by Xaeroxe — APPROVED** [review]
:heart: Thank you! This looks great. I'm going to wait for @ErichDonGubler to approve the C API usage before merging.

File: src/lib.rs (hunk 11 above)
Comment (ErichDonGubler) [reply src/lib.rs:137]: I believe that the laziness of a `Future` or `Stream` is a detail to be defined by whoever's building them and needs to be communicated for each case. Some examples that I think support this understanding:

* Rust's [`std::future::Future` docs state](https://doc.rust-lang.org/std/future/trait.Future.html):

    > A future is a value that may not have finished computing yet.

    To me, this implies work COULD have been done and this COULD have even finished. We just don't know yet.

* [`futures-preview`](https://docs.rs/futures-preview/) (which is intended to become the next stable release of `futures` [has a few examples of leaf `Future`s](https://docs.rs/futures-preview/0.3.0-alpha.19/futures/future/index.html#functions) where the contract is dependent on the semantics of the function called. For example:
    * [`lazy` expects not to do any work until `poll`ed the first time](file:///home/erichdongubler/workspace/futures-rs/target/doc/futures/future/fn.lazy.html).
    * On the flip side, you have functions like `ok` which immediately evaluate to a value. No work is done at all on `poll`ing.

This points leads me to believe that we're responsible for defining the semantics of this `Stream` implementation, so what remains now is arguing why it should be eager or lazy with the PubNub subscription. To me, the answer depends on whether PubNub allows subscription from a specific point in time rather than from the point of subscription. My imagination is that the answer is, "No", based on the fact that PubNub is a real-time platform and probably not interested in maintaining history. But then again, I'm not familiar with PubNub and hope you know the answer.

If PubNub DOESN'T allow checking for publications before point of subscription, my concern with the subscription being lazy is a publication being missed because of other work that may need to be done . For instance, imagine somebody naively implementing a flow in `main` where:

* The user gets a client to PubNub going using `zugzug`. No subscriptions yet.
* User decides on a key they'll use for communication over the pub/sub, and create a subscription stream with that key so as to not miss any messages. However, the subscription hasn't happened yet, unbeknownst to the user!
* User publishes to a well-known key to announce its ad-hoc key to other interested parties.
* Scheduling is opaque and could theoretically wait long enough for a publication to happen on the announced key before the subscription to that key actually happens.

Does that make sense? Anything you see that doesn't sound valid?

Comment (jasonahills) [reply src/lib.rs:137]: That sounds great to me.  We'll keep it as it is then.  Indeed, I'll also get rid of the laziness in the `PublishFuture`.

Comment (jasonahills) [reply src/lib.rs:137]: Actually, I'm going to keep the laziness for `PublishFuture` so that I don't have issues with `notify`-ing outside of a task.

Comment (ErichDonGubler) [reply src/lib.rs:137]: Yeah, I don't see the same concern with publication. Feel free!

**Comment by ErichDonGubler** [comment]
It looks like an audit of C code will take a while -- it looks like client code is responsible for the lifetimes of provided strings to most of the `c-core` API. Eep! Let's do this: I'll create a separate issue for this and we'll get to it...sometime.
